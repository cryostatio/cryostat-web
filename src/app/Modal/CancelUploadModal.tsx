/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { portalRoot } from '@app/utils/utils';
import { Button, Modal } from '@patternfly/react-core';
import * as React from 'react';

export interface CancelUploadModalProps {
  visible: boolean;
  onYes: () => void;
  onNo: () => void;
  title: string;
  message: string;
}

export const CancelUploadModal: React.FC<CancelUploadModalProps> = ({ visible, onYes, onNo, title, message }) => {
  return (
    <Modal
      appendTo={portalRoot}
      width={'40%'}
      isOpen={visible}
      showClose={true}
      onClose={onNo}
      title={title}
      actions={[
        <Button key={'Yes'} variant="primary" onClick={onYes}>
          Yes
        </Button>,
        <Button key={'No'} variant="secondary" onClick={onNo}>
          No
        </Button>,
      ]}
    >
      {message}
    </Modal>
  );
};
