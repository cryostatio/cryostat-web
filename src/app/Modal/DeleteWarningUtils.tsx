/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
export enum DeleteWarningType {
  DeleteActiveRecordings = 'DeleteActiveRecordings',
  DeleteArchivedRecordings = 'DeleteArchivedRecordings',
  DeleteAutomatedRules = 'DeleteAutomatedRules',
  DeleteEventTemplates = 'DeleteEventTemplates',
  DeleteProbeTemplates = 'DeleteProbeTemplates',
  DeleteActiveProbes = 'DeleteActiveProbes',
  DeleteJMXCredentials = 'DeleteJMXCredentials',
  DeleteCustomTargets = 'DeleteCustomTargets',
}

export interface DeleteWarning {
  id: DeleteWarningType;
  title: string;
  label: string;
  description: string;
  ariaLabel: string;
}

export const DeleteActiveRecordings: DeleteWarning = {
  id: DeleteWarningType.DeleteActiveRecordings,
  title: 'Permanently delete Active Recording?',
  label: 'Delete Active Recording',
  description: `Recording and report data will be lost.`,
  ariaLabel: 'Recording delete warning',
};

export const DeleteArchivedRecordings: DeleteWarning = {
  id: DeleteWarningType.DeleteArchivedRecordings,
  title: 'Permanently delete Archived Recording?',
  label: 'Delete Archived Recording',
  description: `Recording and report data will be lost.`,
  ariaLabel: 'Recording delete warning',
};

export const DeleteAutomatedRules: DeleteWarning = {
  id: DeleteWarningType.DeleteAutomatedRules,
  title: 'Permanently delete Automated Rule?',
  label: 'Delete Automated Rule',
  description: `Rule data will be lost.`,
  ariaLabel: 'Automated rule delete warning',
};

export const DeleteEventTemplates: DeleteWarning = {
  id: DeleteWarningType.DeleteEventTemplates,
  title: 'Permanently delete Event Template?',
  label: 'Delete Event Template',
  description: `Custom event template data will be lost.`,
  ariaLabel: 'Event template delete warning',
};

export const DeleteProbeTemplates: DeleteWarning = {
  id: DeleteWarningType.DeleteProbeTemplates,
  title: 'Permanently delete Probe Template?',
  label: 'Delete Probe Template',
  description: `Custom probe template data will be lost.`,
  ariaLabel: 'Probe template delete warning',
};

export const DeleteActiveProbes: DeleteWarning = {
  id: DeleteWarningType.DeleteActiveProbes,
  title: 'Permanently remove Active Probes from the target?',
  label: 'Remove Active Probes',
  description: `Active probes will be removed from the target.`,
  ariaLabel: 'Active Probes remove warning',
};

export const DeleteJMXCredentials: DeleteWarning = {
  id: DeleteWarningType.DeleteJMXCredentials,
  title: 'Permanently delete JMX Credentials?',
  label: 'Delete JMX Credentials',
  description: `Credential data for this target will be lost.`,
  ariaLabel: 'JMX Credentials delete warning',
};

export const DeleteCustomTargets: DeleteWarning = {
  id: DeleteWarningType.DeleteCustomTargets,
  title: 'Permanently delete Custom Target?',
  label: 'Delete Custom Targets',
  description: `Custom target information will be lost.`,
  ariaLabel: 'Custom Targets delete warning',
};

export const DeleteWarningKinds: DeleteWarning[] = [
  DeleteActiveRecordings,
  DeleteArchivedRecordings,
  DeleteAutomatedRules,
  DeleteEventTemplates,
  DeleteProbeTemplates,
  DeleteActiveProbes,
  DeleteJMXCredentials,
  DeleteCustomTargets,
];

export const getFromWarningMap = (warning: DeleteWarningType): DeleteWarning | undefined => {
  const wt = DeleteWarningKinds.find((t) => t.id === warning);
  return wt;
};
