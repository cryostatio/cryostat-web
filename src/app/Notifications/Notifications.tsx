/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import * as React from 'react';
import { BehaviorSubject, Observable } from 'rxjs';
import { concatMap, filter, first, map } from 'rxjs/operators';
import { AlertVariant } from '@patternfly/react-core';
import { nanoid } from 'nanoid';
import { NotificationCategory } from '@app/Shared/Services/NotificationChannel.service';

export interface Notification {
  hidden?: boolean;
  read?: boolean;
  key?: string;
  title: string;
  message?: string | Error;
  category?: string;
  variant: AlertVariant;
  timestamp?: number;
}

export class Notifications {
  private readonly _notifications$: BehaviorSubject<Notification[]> = new BehaviorSubject<Notification[]>([]);
  private readonly _drawerState$: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);

  constructor() {
    this._drawerState$
      .pipe(
        filter((v) => v),
        concatMap(() => this._notifications$.pipe(first()))
      )
      .subscribe((prev) =>
        this._notifications$.next(
          prev.map((n) => ({
            ...n,
            hidden: true,
          }))
        )
      );
  }

  drawerState(): Observable<boolean> {
    return this._drawerState$.asObservable();
  }

  setDrawerState(state: boolean): void {
    this._drawerState$.next(state);
  }

  notify(notification: Notification): void {
    if (!notification.key) {
      notification.key = nanoid();
    }
    notification.read = false;
    if (notification.hidden === undefined) {
      notification.hidden = this._drawerState$.getValue();
    }
    notification.timestamp = +Date.now();
    if (notification.message instanceof Error) {
      notification.message = JSON.stringify(notification.message, Object.getOwnPropertyNames(notification.message));
    } else if (typeof notification.message !== 'string') {
      notification.message = JSON.stringify(notification.message);
    }
    this._notifications$.pipe(first()).subscribe((prev) => {
      prev.unshift(notification);
      this._notifications$.next(prev);
    });
  }

  success(title: string, message?: string | Error, category?: string, hidden?: boolean): void {
    this.notify({ title, message, category, variant: AlertVariant.success, hidden });
  }

  info(title: string, message?: string | Error, category?: string, hidden?: boolean): void {
    this.notify({ title, message, category, variant: AlertVariant.info, hidden });
  }

  warning(title: string, message?: string | Error, category?: string): void {
    this.notify({ title, message, category, variant: AlertVariant.warning });
  }

  danger(title: string, message?: string | Error, category?: string): void {
    this.notify({ title, message, category, variant: AlertVariant.danger });
  }

  notifications(): Observable<Notification[]> {
    return this._notifications$.asObservable();
  }

  unreadNotifications(): Observable<Notification[]> {
    return this.notifications().pipe(map((a) => a.filter((n) => !n.read)));
  }

  actionsNotifications(): Observable<Notification[]> {
    return this.notifications().pipe(map((a) => a.filter((n) => this.isActionNotification(n))));
  }

  cryostatStatusNotifications(): Observable<Notification[]> {
    return this.notifications().pipe(
      map((a) =>
        a.filter(
          (n) => (this.isWsClientActivity(n) || this.isJvmDiscovery(n)) && !Notifications.isProblemNotification(n)
        )
      )
    );
  }

  problemsNotifications(): Observable<Notification[]> {
    return this.notifications().pipe(map((a) => a.filter(Notifications.isProblemNotification)));
  }

  setHidden(key?: string, hidden: boolean = true): void {
    if (!key) {
      return;
    }
    this._notifications$.pipe(first()).subscribe((prev) => {
      for (let n of prev) {
        if (n.key === key) {
          n.hidden = hidden;
        }
      }
      this._notifications$.next(prev);
    });
  }

  setRead(key?: string, read: boolean = true): void {
    if (!key) {
      return;
    }
    this._notifications$.pipe(first()).subscribe((prev) => {
      for (let n of prev) {
        if (n.key === key) {
          n.read = read;
        }
      }
      this._notifications$.next(prev);
    });
  }

  markAllRead(): void {
    this._notifications$.pipe(first()).subscribe((prev) => {
      for (let n of prev) {
        n.read = true;
      }
      this._notifications$.next(prev);
    });
  }

  clearAll(): void {
    this._notifications$.next([]);
  }

  private isActionNotification(n: Notification): boolean {
    return !this.isWsClientActivity(n) && !this.isJvmDiscovery(n) && !Notifications.isProblemNotification(n);
  }

  private isWsClientActivity(n: Notification): boolean {
    return n.category === NotificationCategory.WsClientActivity;
  }

  private isJvmDiscovery(n: Notification): boolean {
    return n.category === NotificationCategory.TargetJvmDiscovery;
  }

  static isProblemNotification(n: Notification): boolean {
    return n.variant === AlertVariant.warning || n.variant === AlertVariant.danger;
  }
}

const NotificationsInstance = new Notifications();

const NotificationsContext = React.createContext(NotificationsInstance);

export { NotificationsContext, NotificationsInstance };
