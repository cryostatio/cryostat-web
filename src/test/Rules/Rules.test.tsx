/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import * as React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { of, Subject } from 'rxjs';
import '@testing-library/jest-dom';
import renderer, { act } from 'react-test-renderer';
import { act as doAct, cleanup, screen, within } from '@testing-library/react';
import { Rules, Rule } from '@app/Rules/Rules';
import { ServiceContext, defaultServices, Services } from '@app/Shared/Services/Services';
import {
  NotificationCategory,
  NotificationChannel,
  NotificationMessage,
} from '@app/Shared/Services/NotificationChannel.service';
import { DeleteAutomatedRules, DeleteWarningType } from '@app/Modal/DeleteWarningUtils';
import { renderWithServiceContextAndRouter } from '../Common';
import { NotificationsContext, NotificationsInstance } from '@app/Notifications/Notifications';

const mockRule: Rule = {
  name: 'mockRule',
  description: 'A mock rule',
  matchExpression: "target.alias == 'io.cryostat.Cryostat' || target.annotations.cryostat['PORT'] == 9091",
  enabled: true,
  eventSpecifier: 'template=Profiling,type=TARGET',
  archivalPeriodSeconds: 0,
  initialDelaySeconds: 0,
  preservedArchives: 0,
  maxAgeSeconds: 0,
  maxSizeBytes: 0,
};
const mockRuleListResponse = { data: { result: [mockRule] as Rule[] } };
const mockRuleListEmptyResponse = { data: { result: [] as Rule[] } };

const mockFileUpload = new File([JSON.stringify(mockRule)], `${mockRule.name}.json`, { type: 'application/json' });
mockFileUpload.text = jest.fn(() => Promise.resolve(JSON.stringify(mockRule)));

const mockDeleteNotification = { message: { ...mockRule } } as NotificationMessage;

const mockUpdateNotification = { message: { ...mockRule, enabled: false } } as NotificationMessage;

const history = createMemoryHistory({ initialEntries: ['/rules'] });

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useRouteMatch: () => ({ url: history.location.pathname }),
  useHistory: () => history,
}));

const downloadSpy = jest.spyOn(defaultServices.api, 'downloadRule').mockReturnValue();
const uploadSpy = jest.spyOn(defaultServices.api, 'uploadRule').mockReturnValue(of(true));
const updateSpy = jest.spyOn(defaultServices.api, 'updateRule').mockReturnValue(of(true));
jest
  .spyOn(defaultServices.api, 'doGet')
  .mockReturnValueOnce(of(mockRuleListEmptyResponse)) // renders correctly empty
  .mockReturnValue(of(mockRuleListResponse));

jest.spyOn(defaultServices.settings, 'deletionDialogsEnabledFor').mockReturnValueOnce(true);

jest
  .spyOn(defaultServices.notificationChannel, 'messages')
  .mockReturnValueOnce(of()) // renders correctly
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // open view to create rules
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // opens upload modal
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // delete a rule when clicked with popup
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // delete a rule when clicked w/o popup
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // remove a rule when receiving notification
  .mockReturnValueOnce(of(mockDeleteNotification))
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // update a rule when receiving notification
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of(mockUpdateNotification))

  .mockReturnValue(of()); // other tests

describe('<Rules />', () => {
  beforeEach(() => {
    history.go(-history.length);
  });

  afterEach(cleanup);

  it('renders correctly', async () => {
    let tree;
    await act(async () => {
      tree = renderer.create(
        <ServiceContext.Provider value={defaultServices}>
          <NotificationsContext.Provider value={NotificationsInstance}>
            <Router location={history.location} history={history}>
              <Rules />
            </Router>
          </NotificationsContext.Provider>
        </ServiceContext.Provider>
      );
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('opens create rule view when Create is clicked', async () => {
    const { user } = renderWithServiceContextAndRouter(<Rules />, { history: history });

    await user.click(screen.getByRole('button', { name: /Create/ }));

    expect(history.entries.map((entry) => entry.pathname)).toStrictEqual(['/rules', '/rules/create']);
  });

  it('opens upload modal when upload icon is clicked', async () => {
    const { user } = renderWithServiceContextAndRouter(<Rules />, { history: history });

    await user.click(screen.getByRole('button', { name: 'Upload' }));

    const modal = await screen.findByRole('dialog');
    expect(modal).toBeInTheDocument();
    expect(modal).toBeVisible();

    const modalTitle = await within(modal).findByText('Upload Automated Rules');
    expect(modalTitle).toBeInTheDocument();
    expect(modalTitle).toBeVisible();

    const dropZoneText = within(modal).getByText('Drag and drop files here');
    expect(dropZoneText).toBeInTheDocument();
    expect(dropZoneText).toBeVisible();
  });

  it('shows a popup when Delete is clicked and then deletes the Rule after clicking confirmation Delete', async () => {
    const { user } = renderWithServiceContextAndRouter(<Rules />, { history: history });

    const deleteRequestSpy = jest.spyOn(defaultServices.api, 'deleteRule').mockReturnValue(of(true));
    const dialogWarningSpy = jest.spyOn(defaultServices.settings, 'setDeletionDialogsEnabledFor');

    await user.click(screen.getByLabelText('Actions'));
    await user.click(await screen.findByText('Delete'));

    expect(screen.getByLabelText(DeleteAutomatedRules.ariaLabel));

    await user.click(screen.getByLabelText("Don't ask me again"));
    await user.click(within(screen.getByLabelText(DeleteAutomatedRules.ariaLabel)).getByText('Delete'));

    expect(deleteRequestSpy).toHaveBeenCalledTimes(1);
    expect(deleteRequestSpy).toBeCalledWith(mockRule.name, true);
    expect(dialogWarningSpy).toBeCalledTimes(1);
    expect(dialogWarningSpy).toBeCalledWith(DeleteWarningType.DeleteAutomatedRules, false);
  });

  it('deletes a rule when Delete is clicked w/o popup warning', async () => {
    const { user } = renderWithServiceContextAndRouter(<Rules />, { history: history });

    const deleteRequestSpy = jest.spyOn(defaultServices.api, 'deleteRule').mockReturnValue(of(true));

    await user.click(screen.getByLabelText('Actions'));
    await user.click(await screen.findByText('Delete'));

    expect(screen.queryByLabelText(DeleteAutomatedRules.ariaLabel)).not.toBeInTheDocument();
    expect(deleteRequestSpy).toHaveBeenCalledTimes(1);
    expect(deleteRequestSpy).toBeCalledWith(mockRule.name, true);
  });

  it('remove a rule when receiving a notification', async () => {
    renderWithServiceContextAndRouter(<Rules />, { history: history });

    expect(screen.queryByText(mockRule.name)).not.toBeInTheDocument();
  });

  it('update a rule when receiving a notification', async () => {
    const subj = new Subject<NotificationMessage>();
    const mockNotifications = {
      messages: (category: string) => (category === NotificationCategory.RuleUpdated ? subj.asObservable() : of()),
    } as NotificationChannel;
    const services: Services = {
      ...defaultServices,
      notificationChannel: mockNotifications,
    };
    const { container } = renderWithServiceContextAndRouter(<Rules />, { history: history, services: services });

    expect(await screen.findByText(mockRule.name)).toBeInTheDocument();

    let labels = container.querySelectorAll('label');
    expect(labels).toHaveLength(1);
    let label = labels[0];
    expect(label).toHaveClass('switch-toggle-true');
    expect(label).not.toHaveClass('switch-toggle-false');

    doAct(() => subj.next(mockUpdateNotification));

    labels = container.querySelectorAll('label');
    expect(labels).toHaveLength(1);
    label = labels[0];
    expect(label).not.toHaveClass('switch-toggle-true');
    expect(label).toHaveClass('switch-toggle-false');
  });

  it('downloads a rule when Download is clicked', async () => {
    const { user } = renderWithServiceContextAndRouter(<Rules />, { history: history });

    await user.click(screen.getByLabelText('Actions'));
    await user.click(await screen.findByText('Download'));

    expect(downloadSpy).toHaveBeenCalledTimes(1);
    expect(downloadSpy).toBeCalledWith(mockRule.name);
  });

  it('updates a rule when the switch is clicked', async () => {
    const { user } = renderWithServiceContextAndRouter(<Rules />, { history: history });

    await user.click(screen.getByRole('checkbox'));

    expect(updateSpy).toHaveBeenCalledTimes(1);
    expect(updateSpy).toBeCalledWith({ ...mockRule, enabled: !mockRule.enabled });
  });

  it('upload a rule file when Submit is clicked', async () => {
    const { user } = renderWithServiceContextAndRouter(<Rules />, { history: history });

    await user.click(screen.getByRole('button', { name: 'Upload' }));

    const modal = await screen.findByRole('dialog');
    expect(modal).toBeInTheDocument();
    expect(modal).toBeVisible();

    const modalTitle = await within(modal).findByText('Upload Automated Rules');
    expect(modalTitle).toBeInTheDocument();
    expect(modalTitle).toBeVisible();

    const dropZoneText = within(modal).getByText('Drag and drop files here');
    expect(dropZoneText).toBeInTheDocument();
    expect(dropZoneText).toBeVisible();

    const uploadButton = within(modal).getByText('Upload');
    expect(uploadButton).toBeInTheDocument();
    expect(uploadButton).toBeVisible();

    const uploadInput = modal.querySelector("input[accept='application/json'][type='file']") as HTMLInputElement;
    expect(uploadInput).toBeInTheDocument();
    expect(uploadInput).not.toBeVisible();

    await user.click(uploadButton);
    await user.upload(uploadInput, mockFileUpload);

    const fileUploadNameText = within(modal).getByText(mockFileUpload.name);
    expect(fileUploadNameText).toBeInTheDocument();
    expect(fileUploadNameText).toBeVisible();

    const submitButton = within(modal).getByText('Submit');
    expect(submitButton).toBeInTheDocument();
    expect(submitButton).toBeVisible();
    expect(submitButton).not.toBeDisabled();

    await user.click(submitButton);

    expect(uploadSpy).toHaveBeenCalled();
    expect(uploadSpy).toHaveBeenCalledWith(mockRule, expect.any(Function), expect.any(Subject));

    expect(within(modal).queryByText('Submit')).not.toBeInTheDocument();
    expect(within(modal).queryByText('Cancel')).not.toBeInTheDocument();

    const closeButton = within(modal).getByText('Close');
    expect(closeButton).toBeInTheDocument();
    expect(closeButton).toBeVisible();
  });
});
