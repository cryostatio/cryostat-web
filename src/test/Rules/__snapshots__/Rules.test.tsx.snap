// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Rules /> renders correctly 1`] = `
<div
  className="pf-v5-c-page__main-group"
>
  <section
    className="pf-v5-c-page__main-section"
  >
    <nav
      aria-label="Breadcrumb"
      className="pf-v5-c-breadcrumb"
      data-ouia-component-id="OUIA-Generated-Breadcrumb-1"
      data-ouia-component-type="PF5/Breadcrumb"
      data-ouia-safe={true}
    >
      <ol
        className="pf-v5-c-breadcrumb__list"
        role="list"
      >
        <li
          className="pf-v5-c-breadcrumb__item"
        >
          <h1
            className="pf-v5-c-breadcrumb__heading"
          >
            Automated Rules
          </h1>
        </li>
      </ol>
    </nav>
    <div
      className="pf-v5-l-stack pf-m-gutter"
    >
      <div
        className="pf-v5-l-stack__item"
      >
        <div
          className="pf-v5-c-card"
          data-ouia-component-id="OUIA-Generated-Card-1"
          data-ouia-component-type="PF5/Card"
          data-ouia-safe={true}
          data-quickstart-id="about-rules"
          id=""
        >
          <div
            className="pf-v5-c-card__title"
          >
            <div
              className="pf-v5-c-card__title-text"
            >
              About Automated Rules
            </div>
          </div>
          <div
            className="pf-v5-c-card__body"
          >
            Automated Rules define a dynamic set of target JVMs to connect to and start
             
            <a
              href="/recordings"
              onClick={[Function]}
            >
              Active Recordings
            </a>
             using a specific 
            <a
              href="/events"
              onClick={[Function]}
            >
              Event Template
            </a>
             
            when the Automated Rule is created and when any new matching target JVMs appear. If your Target JVM connections require JMX Credentials, you can configure these in 
            <a
              href="/security"
              onClick={[Function]}
            >
              Security
            </a>
            . Automated Rules can be configured to periodically copy the contents of the Active Recording to
             
            <a
              href="/archives"
              onClick={[Function]}
            >
              Archives
            </a>
             to ensure you always have up-to-date information about your JVMs.
          </div>
        </div>
      </div>
      <div
        className="pf-v5-l-stack__item"
      >
        <div
          className="pf-v5-c-card"
          data-ouia-component-id="OUIA-Generated-Card-2"
          data-ouia-component-type="PF5/Card"
          data-ouia-safe={true}
          id=""
        >
          <div
            className="pf-v5-c-card__body"
          >
            <div
              className="pf-v5-c-toolbar"
              data-ouia-component-id="OUIA-Generated-Toolbar-1"
              data-ouia-component-type="PF5/Toolbar"
              data-ouia-safe={true}
              id="automated-rule-toolbar"
            >
              <div
                className="pf-v5-c-toolbar__content"
              >
                <div
                  className="pf-v5-c-toolbar__content-section"
                >
                  <div
                    className="pf-v5-c-toolbar__group pf-m-icon-button-group"
                  >
                    <div
                      className="pf-v5-c-toolbar__item"
                    >
                      <button
                        aria-disabled={false}
                        aria-label={null}
                        className="pf-v5-c-button pf-m-primary"
                        data-ouia-component-id="OUIA-Generated-Button-primary-1"
                        data-ouia-component-type="PF5/Button"
                        data-ouia-safe={true}
                        data-quickstart-id="create-rule-btn"
                        disabled={false}
                        onClick={[Function]}
                        role={null}
                        type="button"
                      >
                        Create
                      </button>
                       
                      <button
                        aria-disabled={false}
                        aria-label="Upload"
                        className="pf-v5-c-button pf-m-secondary"
                        data-ouia-component-id="OUIA-Generated-Button-secondary-1"
                        data-ouia-component-type="PF5/Button"
                        data-ouia-safe={true}
                        disabled={false}
                        onClick={[Function]}
                        role={null}
                        type="button"
                      >
                        <svg
                          aria-hidden={true}
                          aria-labelledby={null}
                          className="pf-v5-svg"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 512 512"
                          width="1em"
                        >
                          <path
                            d="M296 384h-80c-13.3 0-24-10.7-24-24V192h-87.7c-17.8 0-26.7-21.5-14.1-34.1L242.3 5.7c7.5-7.5 19.8-7.5 27.3 0l152.2 152.2c12.6 12.6 3.7 34.1-14.1 34.1H320v168c0 13.3-10.7 24-24 24zm216-8v112c0 13.3-10.7 24-24 24H24c-13.3 0-24-10.7-24-24V376c0-13.3 10.7-24 24-24h136v8c0 30.9 25.1 56 56 56h80c30.9 0 56-25.1 56-56v-8h136c13.3 0 24 10.7 24 24zm-124 88c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20zm64 0c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20z"
                          />
                        </svg>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
              <div
                className="pf-v5-c-toolbar__content pf-m-hidden"
                hidden={true}
              >
                <div
                  className="pf-v5-c-toolbar__group"
                />
              </div>
            </div>
            <div
              className="pf-v5-c-empty-state"
            >
              <div
                className="pf-v5-c-empty-state__content"
              >
                <div
                  className="pf-v5-c-empty-state__header"
                >
                  <div
                    className="pf-v5-c-empty-state__icon"
                  >
                    <svg
                      aria-hidden={true}
                      aria-labelledby={null}
                      className="pf-v5-svg"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      viewBox="0 0 512 512"
                      width="1em"
                    >
                      <path
                        d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                      />
                    </svg>
                  </div>
                  <div
                    className="pf-v5-c-empty-state__title"
                  >
                    <h4
                      className="pf-v5-c-empty-state__title-text"
                    >
                      No Automated Rules
                    </h4>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
